### ISSUE TITLE
Host combinator

### ISSUE THREAD

**Issue by alpmestan** [haskell-servant/servant#134]
This one has been discussed in the paper we submitted to WGP if I remember correctly, and I still have the file around so I thought I might as well propose the inclusion of a `Host` combinator to get the host/ip of the client.

``` haskell
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE ScopedTypeVariables #-}
import Network.Socket (SockAddr)
import Network.Wai
import Servant
import Servant.Client

data Host

instance HasServer api => HasServer (Host :> api) where
  type ServerT (Host :> api) m = SockAddr -> ServerT api m

  route Proxy serv req respond =
    route (Proxy :: Proxy api) (serv h) req respond

    where h = remoteHost req

instance HasClient api => HasClient (Host :> api) where
  type Client (Host :> api) = Client api

  clientWithRoute Proxy req baseurl =
    clientWithRoute (Proxy :: Proxy api) req baseurl
```

Nothing fancy really. Request handlers for endpoints which use `Host` get an argument of type `SockAddr` (from `Network.Socket`).

**Comment by jkarni** [comment]
:+1: Can we also add `isSecure`, `vault`, and maybe `httpVersion`? These are all almost exactly the same as above.

**Comment by jkarni** [comment]
oh, and can we name it `RemoteHost` instead of `Host`?

**Comment by alpmestan** [comment]
Yeah, will do. And yes for the renaming too, it's much clearer. Expect a PR very soon =)

**Comment by alpmestan** [comment]
It's called `RemoteHost` and has been merged as part of #141.
